src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesProjectTreeProvider.cs

```diff
@@ -405,118 +413,9 @@ private Task BuildTreeForSnapshotAsync(IDependenciesSnapshot snapshot)
-        /// </param>
-        /// <param name="namedCatalogs">The dictionary of catalogs.</param>
-        /// <returns>The sequence of matching rules.  Hopefully having exactly one element.</returns>
-        private IEnumerable<Rule> GetSchemaForReference(string itemType,
-                                                        bool resolved,
-                                                        IImmutableDictionary<string, IPropertyPagesCatalog> namedCatalogs)
-        {
-            Requires.NotNull(namedCatalogs, nameof(namedCatalogs));
-
-            // Note: usually for default/generic dependencies we have sets of 2 rule schemas:
-            //  - rule for unresolved dependency, they persist in ProjectFile
-            //  - rule for resolved dependency, they persist in ResolvedReference
-            // So to be able to find rule for resolved or unresolved reference we need to be consistent there 
-            // and make sure we do set ResolvedReference persistence for resolved dependencies, since we rely
-            // on that here when pick correct rule schema.
-            // (old code used to check if rule datasource has SourceType=TargetResults, which was true for Resolved,
-            // dependencies. However now we have custom logic for collecting unresolved dependencies too and use 
-            // DesignTime build results there too. That's why we switched to check for persistence).
-            IPropertyPagesCatalog browseObjectCatalog = namedCatalogs[PropertyPageContexts.BrowseObject];
-            IReadOnlyCollection<string> schemas = browseObjectCatalog.GetPropertyPagesSchemas(itemType);
-
-            return from schemaName in browseObjectCatalog.GetPropertyPagesSchemas(itemType)
-                   let schema = browseObjectCatalog.GetSchema(schemaName)
-                   where schema.DataSource != null
-                         && string.Equals(itemType, schema.DataSource.ItemType, StringComparison.OrdinalIgnoreCase)
-                         && (resolved == string.Equals(schema.DataSource.Persistence,
-                                                       RuleDataSourceTypes.PersistenceResolvedReference,
-                                                       StringComparison.OrdinalIgnoreCase))
-                   select schema;
-        }
-
-        private void ApplyProjectTreePropertiesCustomization(
-                        IProjectTreeCustomizablePropertyContext context,
-                        ReferencesProjectTreeCustomizablePropertyValues values)
-        {
-            foreach (IProjectTreePropertiesProvider provider in ProjectTreePropertiesProviders.ExtensionValues())
-            {
-                provider.CalculatePropertyValues(context, values);
-            }
+                    return new TreeUpdateResult(dependenciesNode, lazyFill: false);
+                },
+                GetNextTreeUpdateCancellationToken());
         }
 
```

Comment:
As i mentioned in the github issue, this change is wrong. Before we would have a queue that would schedule next SubmitTreeUpdateAsync only after all previous ones were called. When now OnDependenciesSnapshotChanged can be called after 2 different DT builds/evaluations carrying conflicting changes and if timing will be wrong OnDependenciesSnapshotChanged will be called for them in wrong order which would result in sporadic tree behavior in solution explorer with tree showing wrong dependencies and their states..

Reply:
Snapshot1 (assemblies) and Snapshot2 (assemblies+projects) come very quickly one after another, so quick that OnDependenciesSnapshotChanged executes simultaneously. Which one would first reach line 403?

---
In reply to: [226705432](https://github.com/dotnet/project-system/pull/4143#discussion_r226705432) [](ancestors = 226705432)

Reply:
before i had a lock at the begining of the method that would guaranty first come first serve

---
In reply to: [226714600](https://github.com/dotnet/project-system/pull/4143#discussion_r226714600) [](ancestors = 226714600,226705432)

Reply:
But it's still first come, first served, whether you add the lock or not. Even with a lock, one thread could start, switch out before the lock, and another could race past and get the lock.

Reply:
My point was when two threads execute OnDependenciesSnapshotChanged it could be pretty random who first would hit the line 403 and call CPS SubmitTreeUpdateAsync, because of logic in previous lines in that method event though they are quick. If you feel there will be no timing, its ok try it out. However i would add stress tests to verify snapshots are applied right, otherwise  it would be hard to investigate ven if customers open issues.

---
In reply to: [226736814](https://github.com/dotnet/project-system/pull/4143#discussion_r226736814) [](ancestors = 226736814)

Reply:
I'll follow the data backwards and see whether the code that's firing the dependencies `SnapshotChanged` event is serialising invocations or not.